core/domain/stats_services_test.py

```diff
@@ -551,3 +551,115 @@ def test_record_many_answers_with_preexisting_entry(self):
+
+    def test_exactly_100_answers_returned_if_main_shard_has_100_answers(self):
+        submitted_answer_list = [
+            stats_domain.SubmittedAnswer(
+                'answer a', 'TextInput', 0, 1,
+                exp_domain.EXPLICIT_CLASSIFICATION, {}, 'session_id_v', 10.0)
+        ]
+        # Record 100 answers.
+        stats_services.record_answers(
+            self.EXP_ID, self.exploration.version,
+            self.exploration.init_state_name, 'TextInput',
+            submitted_answer_list * 100)
+
+        sample_answers = stats_services.get_sample_answers(
+            self.EXP_ID, self.exploration.version,
+            self.exploration.init_state_name)
+        self.assertEqual(sample_answers, ['answer a'] * 100)
+
+    def test_all_answers_returned_if_main_shard_has_few_answers(self):
+        submitted_answer_list = [
+            stats_domain.SubmittedAnswer(
+                'answer a', 'TextInput', 0, 1,
+                exp_domain.EXPLICIT_CLASSIFICATION, {}, 'session_id_v', 10.0),
+            stats_domain.SubmittedAnswer(
+                'answer bbbbb', 'TextInput', 1, 0,
+                exp_domain.EXPLICIT_CLASSIFICATION, {}, 'session_id_v', 7.5),
+        ]
+        # Record 2 answers.
+        stats_services.record_answers(
+            self.EXP_ID, self.exploration.version,
+            self.exploration.init_state_name, 'TextInput',
+            submitted_answer_list)
+
+        sample_answers = stats_services.get_sample_answers(
+            self.EXP_ID, self.exploration.version,
+            self.exploration.init_state_name)
+        self.assertEqual(sample_answers, ['answer a', 'answer bbbbb'])
+
+    def test_only_sample_answers_in_main_shard_returned(self):
+        # Use a smaller max answer list size so fewer answers are needed to
+        # exceed a shard.
+        with self.swap(
+            stats_models.StateAnswersModel, '_MAX_ANSWER_LIST_BYTE_SIZE',
+            15000):
+            state_answers = stats_services.get_state_answers(
+                self.EXP_ID, self.exploration.version,
+                self.exploration.init_state_name)
+            self.assertIsNone(state_answers)
+
+            submitted_answer_list = [
+                stats_domain.SubmittedAnswer(
+                    'answer ccc', 'TextInput', 1, 1,
+                    exp_domain.EXPLICIT_CLASSIFICATION, {}, 'session_id_v',
+                    3.0),
+            ]
+            stats_services.record_answers(
+                self.EXP_ID, self.exploration.version,
+                self.exploration.init_state_name, 'TextInput',
+                submitted_answer_list * 100)
+
+        # Verify more than 1 shard was stored. The index shard (shard_id 0)
+        # is not included in the shard count. Since a total of 100 answers were
+        # submitted, there must therefore be fewer than 100 answers in the
+        # index shard.
+        model = stats_models.StateAnswersModel.get('%s:%s:%s:%s' % (
```

Comment:
Can you use ``get_all_models`` instead? I don't think we should expose the format for IDs outside of the model if possible.